### PR TITLE
Add Nek to Travis

### PR DESCRIPTION
**Pull request by jeremylt**
This is as WIP branch to add Nek testing to Travis CI. 

So far I have written up a simple test mode for BP1.

To Do:
- ~~Set up error calculation in BP3~~ done
- ~~Add test mode to BP3~~ done
- ~~Fix BP3 boundary conditions~~ done
- ~~Fix Nek geometry data~~ done
- ~~Fix OKL files~~ done
- ~~Add Nek to 'make prove-all'~~ done
- ~~Add Nek to Travis CI~~ done

I would appreciate feedback as I work on this @thilinarmtb 

This adresses Issue #159

### CODE REVIEW

**Comment by jeremylt** [comment]
I think BP3 is not converging due to boundary conditions. I'm investigating.

**Comment by thilinarmtb** [comment]
I will review this today.

**Comment by thilinarmtb** [comment]
@jeremylt : I had a quick look at this. Can't seem to figure out what is going wrong. I will have more
time to look into this after Wednesday. Is that okay?

**Comment by jeremylt** [comment]
Sounds good to me

**Comment by jeremylt** [comment]
The Nek BPs now run and have an error checking. Next is to add them to the CI and we're set.

**Comment by jeremylt** [comment]
Nek and libCEED seem to be doing different versions of the diffusion operator.

My most recent commit has three versions of Nek5k BP3.

When I use libCEED to compute the true RHS vector and then solve for the true solution vector, I get similar results to PETSc

When I use libCEED's diffusion operator to compute the RHS vector from NEK's true solution vector, I get similar results to the Nek5k benchmark code

When I use Nek5k's diffusion operator to compute the RHS vector from NEK's true solution vector, I get very poor results (Note, this is now the RHS vector is computed in Nek5k's benchmark code)

Results:
 ./run-nek-example.sh -c /cpu/self/ref -e bp3 -n 1 -b 3 -test

 libCEED true RHS
           1 cggo scalar:    41   7.5154E-09   1.0000E-08
  8        8  0.2359E-04   error        41
case scalar:      1  7      8         8          2744      2744  41  3.37668E-02  8.12632E+04  8.23581E-04  3.00139E-07
 libCEED algebraic RHS
           1 cggo scalar:    41   7.5170E-09   1.0000E-08
  8        8  0.3651E-07   error        41
case scalar:      1  7      8         8          2744      2744  41  3.36682E-02  8.15012E+04  8.21176E-04  2.99262E-07
 NEK5000 algebraic RHS
           1 cggo scalar:    59   8.1336E-09   1.0000E-08
  8        8  0.3767E-01   error        59
 ERROR IS TOO LARGE
case scalar:      1  7      8         8          2744      2744  59  4.67726E-02  5.86669E+04  7.92755E-04  2.88905E-07

PETSc Results:
[jeremy@pheonix petsc]$ ./bp3 -degree 7 -local 1800
Global dofs: 3375
Process decomposition: 1 1 1
Local elements: 8 = 2 2 2
Owned dofs: 3375 = 15 15 15
KSP cg CONVERGED_RTOL iterations 52 rnorm 7.417772e-11
Pointwise error (max) 1.560465e-04

**Comment by jeremylt** [comment]
I have attempted to add detJac to the NekBP1, as it currently is not included. This made the BP1 stop converging, so I suspect the difference is in how libCEED is computing derivatives vs how Nek is.

**Comment by jeremylt** [comment]
@jedbrown @thilinarmtb

I found an error in the Nek5k computations of the geometry data that is causing these issues.

If you run BP1, it prints out the det of the Jacobian computed in both libCEED and Nek5k at every quadrature point (I have been using './run-nek-example.sh -c /cpu/self/ref -e bp1 -n 1 -b 0 -test'). libCEED gives a constant value, which is to be expected as the mesh is a simple box mesh. However, Nek5k gives values that bounce around the value computed by libCEED, sometimes not even having one significant digit correct. It looks like a more severe version of single precision round off errors.

I believe that lines 450-457 have an error. Past that I am at a bit of a loss, as dxmq (the colocated derivative matrix) seems to be correct, intp_restd seems to be correct, and loc_grad3 is a minor modification on a built in Nek5k library function.

Of note, when I run the benchmark repo version of this code, I find the same incorrect values of det Jac being computed.

**Comment by jeremylt** [comment]
I think I got it. It was an issue with using a global array that was not as large that was asserted in the subroutine.

**Comment by jeremylt** [comment]
I caught a bug in my implimentation of multifield identity restrictions in OCCA. The fix has been included here.

**Comment by jeremylt** [comment]
@jedbrown @thilinarmtb 

We now have working Nek5K+libCEED BPs (1 and 3) and integration of Nek5k into Travis CI.

Of note - currently the Nek BPs are only being run on the linux CI, not OSX, as Nek's 'maketools' is not playing nice with gcc 8, which is needed in the OSX environment for the rull range of dependencies for 'prove-all'. I put up a Nek5k issue for it: https://github.com/Nek5000/Nek5000/issues/563

File: .travis.yml [hunk 1]
```diff
@@ -86,13 +86,15 @@ install:
         && cd petsc-3.10.0
         && ./configure --with-debugging=0 COPTFLAGS=-O --with-mpi=0 --with-fc=0 --with-cxx=0 --prefix=$PETSC_INSTALL
         && make
-        && make install )
+        && make install
+        && cd .. )
```
Comment (jedbrown) [inline .travis.yml]: The `( list )` syntax runs in a sub-shell. The `cd ..` shouldn't be needed. Why was this added?

Can we use a sub-shell below as well?

Comment (jeremylt) [reply .travis.yml]: Just a mistake from my ignorance of the sub shell syntax , fixed up now.

File: Makefile [hunk 2]
```diff
@@ -288,9 +297,12 @@ prove : $(tests) $(examples)
 prv : ;@$(MAKE) $(MFLAGS) V=$(V) prove
 
 alltests := $(tests) $(examples) $(if $(MFEM_DIR),$(mfemexamples)) $(if $(PETSC_DIR),$(petscexamples))
-prove-all : $(alltests)
+fulltestlist = $(alltests) $(if $(NEK5K_DIR), $(nekexamples))
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply Makefile]: Because Nek is built from a bash script rather than in the same way the MFEM and PETSc examples built, I've had trouble making a single Makefile command both provide the commands to build the example and provide the list of files in the `build` directory to run `tap.sh` against. I can investigate further.

**Comment by jeremylt** [comment]
I have to admit, I am a bit confused by the error message you got.
```
cd nek5000 && ./make-nek-tests.sh
Building example: bp1 ...
make[2]: Entering directory '/home/jed/src/libceed/examples/nek5000'
make[2]: *** No targets specified and no makefile found.  Stop.
make[2]: Leaving directory '/home/jed/src/libceed/examples/nek5000'
```
is the message that `make nek` gives from the Makefile in `examples` in the current Master (and thus in my branch). I think it is an artifact of how the Nek build system works. I think this has worked this way for as long as we have had the Nek5k BPs in libCEED.

The Makefile is looking for `NEK5K_DIR` to be set or for Nek5000 to be in a sibbling directory. The message
```
./makenek: line 69: cd: /home/jed/Nek5000: No such file or directory
./makenek: line 73: /home/jed/src/libceed/examples/nek5000/core/makenek.inc: No such file or directory
  Building bp1 failed. See bp1.build.log for details.
```
tells me that `NEK5K_DIR` got muddled some where along the line after copying `makenek`? I'll try to reproduce.

**Comment by jedbrown** [comment]
The repetition is sad and error-prone, though I'm not sure what to do about it, short of Nek modernizing. Anyway, the issue appears to be in `makenek`

```
: ${NEK_SOURCE_ROOT:="$HOME/Nek5000"}
```
which ignores the `NEK5K_DIR` that you've been painstakingly repeating in every file.

```
$ make MAGMA_DIR= NEK5K_DIR=$HOME/src/Nek5000 prove-all
make: 'lib' with optional backends: /cpu/occa /gpu/occa /omp/occa
(export CC FC && cd examples && make prepnektests)
make[1]: Entering directory '/home/jed/src/libceed/examples'
cd nek5000 && ./make-nek-tests.sh
Building example: bp1 ...
make[2]: Entering directory '/home/jed/src/libceed/examples/nek5000'
make[2]: *** No targets specified and no makefile found.  Stop.
make[2]: Leaving directory '/home/jed/src/libceed/examples/nek5000'
./makenek: line 69: cd: /home/jed/Nek5000: No such file or directory
./makenek: line 73: /home/jed/src/libceed/examples/nek5000/core/makenek.inc: No such file or directory
  Building bp1 failed. See bp1.build.log for details.
```

**Comment by jedbrown** [comment]
I think you'd have to export `NEK_SOURCE_ROOT="$NEK5K_DIR"`. Can you try using different paths on your machine?

**Comment by jeremylt** [comment]
The top of my `makenek` is slightly different than yours:

```
# assign version tag
mver=17.0.4

# set defaults
: ${SOURCE_ROOT:="$HOME/Nek5000"}
: ${FC:="mpif77"}
: ${CC:="mpicc"}
```

The `make-nek-examples.sh` file sets `SOURCE_ROOT` but does not set `NEK_SOURCE_ROOT`.

I agree that this series of delicate scripts is not ideal, but trying to put Makefiles in the Nek examples folder lead to conflicts with Nek's build system, so I don't know that we can do much without Nek changing their build system.

As a side note, there are so many scripts interfering with themselves that `make-nek-examples.sh` is failing to correctly handle simple output redirections (see lines 2446 to 2541 which should be writing to file: https://travis-ci.org/CEED/libCEED/jobs/472267278).

**Comment by jedbrown** [comment]
https://github.com/Nek5000/Nek5000/commit/16bfa9e25b60c745b904ad09c1d5ab9cf1bd3d0c#diff-c25dfaf7586e55b4477ce59b0d652e5b

I could probably figure out the redirect issue if you can reduce it somewhat, but this tax should really be paid by Nek developers.

**Comment by jeremylt** [comment]
I have modifed some of the scripts so that everything runs cleanly. The only open question to me is if my scripts interfacing with the Nek build system can be improved to be less fragile.

**Comment by jeremylt** [comment]
@thilinarmtb, if you get a chance, I'd appreciate your review on this.

**Comment by jeremylt** [comment]
@YohannDudouit, I adapted this PR to be compatible with the CUDA backend. Can you please check it? (all of the relevant work should be in the most recent commit)

@jedbrown, I think we're finally ready once @YohannDudouit says it works with CUDA.

File: examples/nek5000/bp1.cu [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+// *****************************************************************************
+extern "C" __global__ void Setup(void *ctx, CeedInt Q,
+                                 Fields_Cuda fields) {
+  CeedScalar *rho = fields.outputs[0], *rhs = fields.outputs[1];
+  const CeedScalar *x = (const CeedScalar *)fields.inputs[0];
+  const CeedScalar *J = (const CeedScalar *)fields.inputs[1];
+  const CeedScalar *w = fields.inputs[2];
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply examples/nek5000/bp1.cu]: I imagine it was just an oversight - I copied it from the PETSc bp1.cu

[the remark this answers is not in the document]
Comment (jeremylt) [reply examples/nek5000/bp1.cu]: Updated
